src/radical/pilot/agent/lm/spark.py

```diff
@@ -0,0 +1,285 @@
+                node_name = lrms.node_list[0]
+            else:
+                logger.info("Bootstrap SPARK on " + socket.gethostname())
+                node_name =  socket.gethostname()
+            
+            VERSION = "1.5.2"
+            SPARK_DOWNLOAD_URL= "http://d3kbcqa49mib13.cloudfront.net/spark-1.5.2-bin-hadoop2.6.tgz" #prebuilt
+            #Download the tar file
+            opener = urllib.FancyURLopener({})
+            download_destination = os.path.join(os.getcwd(),"spark-" + VERSION + ".tar.gz")
+            logger.info("Download: %s to %s"%(SPARK_DOWNLOAD_URL, download_destination))
+            opener.retrieve(SPARK_DOWNLOAD_URL, download_destination)
+            spark_tar = "spark-" + VERSION + ".tar.gz"
+            if not os.path.isfile(spark_tar):
+                raise RuntimeError("Spark wasn't downloaded properly. Please try again")
+
+            os.system("tar -xzf" + spark_tar + "; rm " + spark_tar ) #untar and delete tarball 
+            os.system("mv spark-1.5.2-bin-hadoop2.6 spark-1.5.2")
+            spark_home = os.getcwd() + '/spark-' + VERSION
+
+            #-------------------------------------------------------------------
+            platform_os = sys.platform
+            java_home = os.environ.get('JAVA_HOME')
+
+            if platform_os == "linux" or platform_os == "linux2":
+                if not java_home:
+                    java = ru.which('java')
+                    if java != '/usr/bin/java':
+                        jpos=java.split('bin')
+                    else:
+                        jpos = os.path.realpath('/usr/bin/java').split('bin')
+
+                    if jpos[0].find('jre') != -1:
+                        java_home = jpos[0][:jpos[0].find('jre')]
+                    else:
+                        java_home = jpos[0]       
+            else:
+                if not java_home:
+                    try:
+                        java_home = subprocess.check_output("/usr/libexec/java_home").split()[0]
+                    except Exception:
+                        java_home = '/Library/Java/Home'
+
+            # if no installation found install scala 2.10.4
+            scala_home=ru.which('scala')
+            if not scala_home:
+                os.system('cd')
+                os.system('wget http://www.scala-lang.org/files/archive/scala-2.10.4.tgz')
+                if not os.path.isfile('scala-2.10.4.tgz'):
+                    raise RuntimeError("Scala wasn't downloaded properly. Please try again")
+                os.system('tar -xvf scala-2.10.4.tgz ; cd scala-2.10.4 ; export PATH=`pwd`/bin:$PATH; export SCALA_HOME=`pwd`')
+                os.system('rm scala-2.10.4.tgz')
+                scala_home = os.getcwd() + '/scala-2.10.4'
+                os.system('cd')
+
+
+            if lrms.node_list[0]!='localhost':
+                hostname = subprocess.check_output('/bin/hostname').split(lrms.node_list[0])[1].split('\n')[0]
+
```

Comment:
Ah, so its not a `hostname` you are getting here, but a domain name?  You may want to rename that then.  Ar you sure that a node name is never fully qualified?  Because what you do there will break then.  Also, are you sure that the output of `/bin/hostname` *is* fully qualified?  Because it often isn't ;)

```
$ ssh archer hostname
$ ssh archer hostname
eslogin003
```

Lastly, the domain returned on the headnode might not be the domain used by the compute nodes, eg. if multiple network cards are involved, which is often the case on clusters.  Guessing host names like this is tricky...

Would spark be able to handle IP addresses?  If so, you may want to look up the addresses for the nodes and use those.  That still can go wrong depending on network configurations, but does not suffer from the domain ambiguity and parsing problems....